apps/aeutils/src/aeu_mp_trees.erl

```diff
@@ -0,0 +1,543 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc Merkle Patricia Trees
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(aeu_mp_trees).
+
+-export([ new/0
+        , new/2
+        , delete/2
+        , get/2
+        , pp/1
+        , put/3
+        , root_hash/1
+        ]).
+
+-export_type([tree/0
+             ]).
+
+-record(mpt, { hash = <<>> :: <<>> | hash() %% <<>> for the empty tree
+             , db   = dict:new() :: dict:dict()
+             }).
+
+-opaque tree() :: #mpt{}.
+
+-type tree_node() :: null() | leaf() | extension() | branch().
+
+-type null()         :: <<>>.
+-type leaf()         :: {'leaf', path(), value()}.
+-type extension()    :: {'ext', path(), hash()}.
+-type branch()       :: {'branch', branch_tuple()}.
+-type branch_tuple() :: {enc_node(), enc_node(), enc_node(), enc_node(),
+                         enc_node(), enc_node(), enc_node(), enc_node(),
+                         enc_node(), enc_node(), enc_node(), enc_node(),
+                         enc_node(), enc_node(), enc_node(), enc_node(),
+                         value()}.
+
+-type enc_node()       :: null() | short_rlp_node() | hash().
+-type short_rlp_node() :: aeu_rlp:encoded(). %% size < 32 bytes
+
+-type raw_node()      :: null() | raw_leaf() | raw_extension() | raw_branch().
+-type raw_leaf()      :: [encoded_path()|value()]. %% [encoded_path(), value()].
+-type raw_extension() :: [encoded_path()|hash()].  %% [encoded_path(), hash()].
+-type raw_branch()    :: [enc_node()|value()].     %% [16x enc_node() + value()]
+
+-type path()         :: <<>> | key(). %% Hex string.
+-type encoded_path() :: binary().  %% Compact encoding of hex sequence.
+
+-type key()          :: <<_:4,_:_*4>>. %% Non-empty hexstring
+-type value()        :: aeu_rlp:encodable().
+-type hash()         :: <<_:256>>.
+
+-type db()           :: dict:dict().
+
+%%-define(debug_on, true).
+-ifdef(debug_on).
```

Comment:
the default macro is DEBUG

Reply:
I can remove the debug stuff altogether.